citizensadvice/design-system
# Generate examples tasks from component previews

## Background

As of https://github.com/citizensadvice/design-system/pull/1076 we've introduced an experimental rails engine based on view components as a future replacement for our standalone haml partials.

Currently we use a custom webpack loader to let us use our haml partials from storybook. This _works_, but requires us to recreate parts of rails on an ad-hoc basis. With the new engine we have a full copy of rails available to us so we'd like to use that somehow to generate examples for the docs.

## Goals

- Provide a way of generating plain HTML examples from component previews.
- Use at least one generated example to replace an existing example within the docs.
- The backstop tests should pass with little or no modification.

## Changes

### Adds a task for generating examples

Adds a rake task that generates plain html examples from component previews which can then be used in storybook.

```
./bin/generate-examples
```

I've swapped out the pagination example—which was already plain html—with one generated by this new task. If the backstop tests pass then I've managed to swap out a visually like-for-like example.

I've also added an example documentation section showing how the component is used from within rails. It's felt OK hand-writing the usage docs for the existing partials, so have followed that approach here too rather than trying to generate anything.

### Adds a watcher to rebuild examples on change

Adds a version of the start command which also runs a watcher which rebuilds the examples when previews are updated; which in turn triggers a storybook update giving us live-reload behaviour.

```
npm run start:watch
```

We might eventually want to make this the default for `npm start`, but for now I've kept it as a separate command.

I went with using npm run tasks to do this rather than ruby scripts as all storybook related scripts are already using node.

## How to test these changes

- Run storybook using `npm run start:watch`
- Go to `http://localhost:6006/?path=/story/components-pagination--example`
- Try editing the parameters for `with_fifth_page` in `pagination_component_preview.rb` and confirm the example live reloads (tip: try reducing `num_pages` to < 10)
- Add a new component preview example and run `./bin/generate-examples` manually and confirm you get the expected output

## Other options considered

I did look at https://github.com/jonspalmer/view_component_storybook which looks like a great if you want really tight integration with Storybook. For our purposes we want _less_ integration, with a longer term goal of building our own docs using the component library itself. So for that reason I went for a more lo-fi approach.

💭 Thoughts very welcome as I'm sure there's plenty of different ways we could go about this.
